python-frontend/src/test/java/org/sonar/python/lexer/PythonLexerTest.java

```diff
@@ -146,28 +155,254 @@ void longbytes_literal() {
+  }
+
+  @Test
+  void fstring_with_escaped_braces() {
+    assertThat(fStringLexer.lex("f\"abc{{a}} { b + 3}\""), allOf(
+      hasToken("f\"", PythonTokenType.FSTRING_START),
+      hasToken("abc{a} ", PythonTokenType.FSTRING_MIDDLE),
+      hasToken("{", PythonPunctuator.LCURLYBRACE),
+      hasToken("b", GenericTokenType.IDENTIFIER),
+      hasToken("+", PythonPunctuator.PLUS),
+      hasToken("3", PythonTokenType.NUMBER),
+      hasToken("}", PythonPunctuator.RCURLYBRACE),
+      hasToken("\"", PythonTokenType.FSTRING_END)));
+  }
+
+  @Test
+  void fstring_format_specifier() {
+    assertThat(fStringLexer.lex("f\"abc {a + b:.3f}\""), allOf(
+      hasToken("f\"", PythonTokenType.FSTRING_START),
+      hasToken("abc ", PythonTokenType.FSTRING_MIDDLE),
+      hasToken("{", PythonPunctuator.LCURLYBRACE),
+      hasToken("a", GenericTokenType.IDENTIFIER),
+      hasToken("+", PythonPunctuator.PLUS),
+      hasToken("b", GenericTokenType.IDENTIFIER),
+      hasToken(":", PythonPunctuator.COLON),
+      hasToken(".3f", PythonTokenType.FSTRING_MIDDLE),
+      hasToken("}", PythonPunctuator.RCURLYBRACE),
+      hasToken("\"", PythonTokenType.FSTRING_END)));
+  }
+
+  @Test
+  void fstring_nested_fields_format_specifier() {
+    assertThat(fStringLexer.lex("f\"abc {a + b:{width}.{length}}\""), allOf(
+      hasToken("f\"", PythonTokenType.FSTRING_START),
+      hasToken("abc ", PythonTokenType.FSTRING_MIDDLE),
+      hasToken("{", PythonPunctuator.LCURLYBRACE),
+      hasToken("a", GenericTokenType.IDENTIFIER),
+      hasToken("+", PythonPunctuator.PLUS),
+      hasToken("b", GenericTokenType.IDENTIFIER),
+      hasToken(":", PythonPunctuator.COLON),
+      hasToken("{", PythonPunctuator.LCURLYBRACE),
+      hasToken("width", GenericTokenType.IDENTIFIER),
+      hasToken("}", PythonPunctuator.RCURLYBRACE),
+      hasToken("{", PythonPunctuator.LCURLYBRACE),
+      hasToken("length", GenericTokenType.IDENTIFIER),
+      hasToken("}", PythonPunctuator.RCURLYBRACE),
+      hasToken("}", PythonPunctuator.RCURLYBRACE),
+      hasToken("\"", PythonTokenType.FSTRING_END)));
+  }
+
+  @Test
+  void fstring_date_format_specifier() {
+    assertThat(fStringLexer.lex("f\"{date:%B %d, %Y}\""), allOf(
+      hasToken("f\"", PythonTokenType.FSTRING_START),
+      hasToken("{", PythonPunctuator.LCURLYBRACE),
+      hasToken("date", GenericTokenType.IDENTIFIER),
+      hasToken(":", PythonPunctuator.COLON),
+      hasToken("%B %d, %Y", PythonTokenType.FSTRING_MIDDLE),
+      hasToken("}", PythonPunctuator.RCURLYBRACE),
+      hasToken("\"", PythonTokenType.FSTRING_END)));
+  }
+
+  @Test
+  void fstring_complex_format_specifier() {
+    assertThat(fStringLexer.lex("f\"{line = !r:20}\""), allOf(
+      hasToken("f\"", PythonTokenType.FSTRING_START),
+      hasToken("{", PythonPunctuator.LCURLYBRACE),
+      hasToken("line", GenericTokenType.IDENTIFIER),
+      hasToken("=", PythonPunctuator.ASSIGN),
+      hasToken("!", GenericTokenType.UNKNOWN_CHAR),
```

Comment:
Not 100% sure about this `UNKNOWN_CHAR` I did not find how I should tokenize `!r`

Reply:
Yes there a no specific token and it seems to work with the grammar with the `unknown char`